pyo3-ffi/src/object.rs

```diff
@@ -72,99 +104,85 @@ pub unsafe fn Py_TYPE(ob: *mut PyObject) -> *mut PyTypeObject {
     (*ob).ob_type
 }
 
+// PyLong_Type defined in longobject.rs
+// PyBool_Type defined in boolobject.rs
+
 #[inline]
 pub unsafe fn Py_SIZE(ob: *mut PyObject) -> Py_ssize_t {
-    (*(ob as *mut PyVarObject)).ob_size
+    debug_assert_ne!((*ob).ob_type, std::ptr::addr_of_mut!(crate::PyLong_Type));
+    debug_assert_ne!((*ob).ob_type, std::ptr::addr_of_mut!(crate::PyBool_Type));
+    (*ob.cast::<PyVarObject>()).ob_size
 }
 
 #[inline]
 pub unsafe fn Py_IS_TYPE(ob: *mut PyObject, tp: *mut PyTypeObject) -> c_int {
     (Py_TYPE(ob) == tp) as c_int
 }
 
+#[inline(always)]
+#[cfg(all(Py_3_12, not(target_pointer_width = "32")))]
+pub unsafe fn _Py_IsImmortal(op: *mut PyObject) -> c_int {
+    (((*op).ob_refcnt.ob_refcnt as crate::PY_INT32_T) < 0) as c_int
```

Comment:
Couldn't this use the split representation? Or is this copied verbatim from the C header?

Reply:
Yep copied verbatim: https://github.com/python/cpython/blob/e3b5ed7b1cb8ad12aa285cf78303c6dda0953181/Include/object.h#L243